apps/omg/test/omg/state/core_test.exs

```diff
@@ -49,6 +49,141 @@ defmodule OMG.State.CoreTest do
+      tx = create_recovered([{1, 0, 0, alice}], @eth, [{alice, 10}])
+
+      state
+      |> Core.with_utxos(%{})
+      |> Core.exec(tx, :no_fees_required)
+      |> fail?(:utxo_not_found)
+    end
+
+    @tag fixtures: [:alice, :bob, :state_empty]
+    test "all transaction inputs are merged from db", %{alice: alice, bob: bob, state_empty: state} do
+      tx = create_recovered([{1000, 0, 0, alice}, {1000, 1, 0, alice}], @eth, [{bob, 7}, {alice, 3}])
+
+      db_utxos = make_utxos([{1000, 0, 0, alice, @eth, 5}, {1000, 1, 0, alice, @eth, 5}])
+
+      state
+      |> Core.with_utxos(db_utxos)
+      |> Core.exec(tx, :no_fees_required)
+      |> success?()
+    end
+
+    @tag fixtures: [:alice, :bob, :state_empty]
+    test "transaction utxos are mixed in memory and db", %{alice: alice, bob: bob, state_empty: state} do
+      tx = create_recovered([{1000, 0, 0, alice}, {1, 0, 0, alice}], @eth, [{bob, 7}, {alice, 3}])
+
+      db_utxos = make_utxos([{1000, 0, 0, alice, @eth, 8}])
+
+      state
+      |> do_deposit(alice, %{amount: 2, currency: @eth, blknum: 1})
+      |> Core.with_utxos(db_utxos)
+      |> Core.exec(tx, :no_fees_required)
+      |> success?()
+    end
+
+    @tag fixtures: [:alice, :bob, :state_alice_deposit]
+    test "spending utxo that resist in memory - double spend impossible",
+         %{alice: alice, bob: bob, state_alice_deposit: state} do
+      tx = create_recovered([{1, 0, 0, alice}], @eth, [{bob, 7}, {alice, 3}])
+
+      state
+      |> Core.exec(tx, :no_fees_required)
+      |> success?()
+      |> Core.exec(tx, :no_fees_required)
+      |> fail?(:utxo_not_found)
+    end
+
+    @tag fixtures: [:state_alice_deposit]
+    test "extending state with same utxos does not change it", %{state_alice_deposit: state} do
+      db_utxos = state.utxos
+
+      state
+      |> Core.with_utxos(db_utxos)
+      |> same?(state)
+    end
+
+    @tag fixtures: [:alice, :bob, :state_empty]
+    test "extending state partially", %{alice: alice, bob: bob, state_empty: state} do
+      fst_utxo = make_utxos([{1000, 0, 0, alice, @eth, 6}])
+      snd_utxo = make_utxos([{1000, 5, 0, alice, @eth, 6}])
+
+      tx = create_recovered([{1000, 0, 0, alice}, {1000, 5, 0, alice}], @eth, [{bob, 10}])
+
+      state
+      |> Core.with_utxos(fst_utxo)
+      |> Core.exec(tx, :no_fees_required)
+      |> fail?(:utxo_not_found)
+      |> Core.with_utxos(snd_utxo)
+      |> Core.exec(tx, :no_fees_required)
+      |> success?()
+    end
+
+    @tag fixtures: [:alice, :bob, :state_empty]
+    test "recent spends are stored regardless state extension", %{alice: alice, bob: bob, state_empty: state} do
+      init_state = do_deposit(state, alice, %{amount: 10, currency: @eth, blknum: 1})
+      tx1 = create_recovered([{1, 0, 0, alice}], @eth, [{bob, 10}])
+
+      {:ok, _, state1} = Core.exec(init_state, tx1, :no_fees_required)
+
+      %Core{recently_spent: spent1} = state1
```

Comment:
now this is a bit awkward... why are we constantly adding cases where we test via internals? here and next `test`.

I'm guessing, you need to transform the tests into more behavior-driven - start with the name, the name of the test is coupled to internals. What is the behavior that we're after here?

Reply:
I drop these tests as this is an detail but we actually don't have to clear `recently_spent` on form_block